c/tskit/tables.c

```diff
@@ -5341,6 +5334,547 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+            tail->right = right;
+        } else {
+            x = ibd_finder_alloc_segment(self, left, right, output_id);
+            if (x == NULL) {
+                ret = TSK_ERR_NO_MEMORY;
+                goto out;
+            }
+            tail->next = x;
+            self->ancestor_map_tail[input_id] = x;
+        }
+    }
+out:
+    return ret;
+}
+
+// Define functions that initialise the ibd_finder from a table collection/
+// (for now -- need number of samples, which comes from tree seq.)
+static int
+ibd_finder_init_samples(tsk_ibd_finder_t *self, tsk_id_t *samples)
+{
+    int ret = 0;
+    size_t j;
+
+    /* Go through the samples to check for errors. */
+    for (j = 0; j < self->num_samples; j++) {
+        if (samples[j] < 0 || samples[j] > (tsk_id_t) self->tables->nodes.num_rows) {
+            ret = TSK_ERR_NODE_OUT_OF_BOUNDS;
+            goto out;
+        }
+        if (self->is_sample[samples[j]]) {
+            ret = TSK_ERR_DUPLICATE_SAMPLE;
+            goto out;
+        }
+        self->is_sample[samples[j]] = true;
+        ret = ibd_finder_add_ancestry(
+            self, samples[j], 0, self->tables->sequence_length, samples[j]);
+        if (ret != 0) {
+            goto out;
+        }
+
+        // Initialise the sample id map.
+        self->sample_id_map[samples[j]] = j;
+    }
+out:
+    return ret;
+}
+
+static void
+ibd_finder_init_oldest_parents(tsk_ibd_finder_t *self)
+{
+    int j, c;
+    const tsk_edge_table_t *input_edges = &self->tables->edges;
+    size_t num_edges = input_edges->num_rows;
+    double *node_times = self->tables->nodes.time;
+    int *p;
+
+    // First, initialise all values to their default (the same node)
+    for (j = 0; j < (int) self->tables->nodes.num_rows; j++) {
+        self->oldest_parent[j] = j;
+    }
+
+    // Replace the default values if an older parent node is found.
+    if (num_edges > 0) {
+        for (j = 0; j < (int) num_edges; j++) {
+            c = input_edges->child[j];
+            p = &self->oldest_parent[c];
+            if (node_times[*p] < node_times[(int) input_edges->parent[j]]) {
+                self->oldest_parent[c] = input_edges->parent[j];
+            }
+            assert(&self->oldest_parent[c] != NULL);
+        }
+    }
+}
+
+static int
+ibd_finder_init(tsk_ibd_finder_t *self, double sequence_length, tsk_id_t *samples,
+    size_t num_samples, tsk_table_collection_t *tables, double min_length,
+    double max_time)
+{
+    int ret = 0;
+    size_t num_nodes_alloc;
+
+    memset(self, 0, sizeof(tsk_ibd_finder_t));
+    self->samples = samples;
+    self->num_samples = num_samples;
+    self->sequence_length = sequence_length;
+    self->num_nodes = tables->nodes.num_rows;
+    self->tables = tables;
+    self->num_pairs = num_samples * (num_samples - 1) / 2;
+    self->min_length = min_length;
+    self->max_time = max_time;
+    if (max_time == 0) {
+        self->max_time = 2 * self->tables->nodes.time[self->num_nodes - 1];
```

Comment:
This seems a bit arbitrary - how about DBL_MAX? This would be a good default, I think.

Reply:
My thought where that (assuming the nodes are ordered by backwards-time), `self->tables->nodes.time[self->num_nodes - 1]` should be the time of the oldest node in the tree, so setting `max_time` to be twice this should safely ensure it is never exceeded. I'm happy to change it to DBL_MAX though!